src/Interpreters/Cache/Metadata.cpp

```diff
@@ -272,9 +271,17 @@ void CacheMetadata::doCleanup()
         }
         catch (...)
         {
-            tryLogCurrentException(__PRETTY_FUNCTION__);
+            LOG_ERROR(log, "Error while removing key {}: {}", cleanup_key, getCurrentExceptionMessage(false));
             chassert(false);
         }
+
+        /// Remove key from metadata AFTER deleting key directory, because otherwise key lock is
```

Comment:
I didn't understand how moving `erase(it)` help here. The key lock is held until the end of the loop iteration. So, it looks the comment is not correct. Did I miss something?

Reply:
> The key lock is held until the end of the loop iteration

it is, but `KeyGuard` object lies inside `metadata[key]`, so if we do `erase(it)` the lock is deleted as well (so the last shared pointer to this key_metadata is hold by LockedKey which holds the lock in this iteration), which means that the thread which will come to cache to take the same key - it will create a new key metadata and a new KeyGuard will be used.

Reply:
AFAIS, `erase(it)` removes the `KeyMetadataPtr` from cache metadata, but the iterator is still holding the corresponding shared pointer on `KeyMetadata`, so the `KeyMetadata` destruction should happen at the end of the loop. 
BTW, It'd be better to make it more explicit, by assigning `KeyMetadataPtr` in `it->second` to a local variable